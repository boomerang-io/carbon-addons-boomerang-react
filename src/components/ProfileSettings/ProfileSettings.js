import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import {
  Button,
  Checkbox,
  StructuredListWrapper,
  StructuredListHead,
  StructuredListBody,
  StructuredListRow,
  StructuredListCell,
  StructuredListSkeleton,
} from 'carbon-components-react';
import {
  ModalHeader,
  ModalBody,
  ModalFooter,
} from 'carbon-components-react/lib/components/ComposedModal';
import { settings } from 'carbon-components';
import ErrorMessage from '../ErrorMessage';
import HeaderMenuUser from '../HeaderMenuUser';
import notify from '../Notifications/notify';
import ToastNotification from '../Notifications/ToastNotification';
import sortBy from 'lodash/sortBy';

const { prefix } = settings;

ProfileSettings.propTypes = {
  baseServiceUrl: PropTypes.string.isRequired,
};

function ProfileSettings({ baseServiceUrl, src, userName }) {
  const [teams, setTeams] = useState([]);
  const [isLoading, setIsLoading] = useState();
  const [isSubmitting, setIsSubmitting] = useState();
  const [isLoadingError, setIsLoadingError] = useState();
  const [isSubmitError, setIsSubmitError] = useState();

  useEffect(() => {
    setIsLoading(true);
    axios(`${baseServiceUrl}/launchpad/users`)
      .then((response) => setTeams(response.data.lowerLevelGroups))
      .catch((err) => setIsLoadingError(err))
      .then(() => {
        setIsLoading(false);
      });
  }, [baseServiceUrl]);

  function handleSubmit({ closeModal }) {
    setIsSubmitting(true);
    console.log(teams);
    axios
      .patch(`${baseServiceUrl}/users/profile`, {
        lowerLevelGroups: teams,
      })
      .then(() => {
        setIsSubmitting(false);
        notify(
          <ToastNotification
            subtitle="Successfully update user settings"
            title="Update Settings"
            kind="success"
          />,
          { containerId: `${prefix}--bmrg-header-notifications` }
        );
        // want to clear out any error state
        setIsSubmitError(false);
        closeModal();
      })
      .catch((err) => {
        setIsSubmitting(false);
        setIsSubmitError(true);
      });
  }

  const visibleTeamCount = teams.filter((team) => team.visible).length;
  const allTeamsAreChecked = teams.length === visibleTeamCount;
  const someTeamsAreChecked = visibleTeamCount > 0 && !allTeamsAreChecked;

  function batchChangeTeamVisibility(action) {
    const updatedTeams = teams.map((team) => ({ ...team, visible: !allTeamsAreChecked }));
    setTeams(updatedTeams);
  }

  function handleUpdateTeamVisibility(id, isVisible) {
    const newTeams = [...teams];
    for (let team of newTeams) {
      if (team.id === id) {
        team.visible = isVisible;
      }
    }
    setTeams(newTeams);
  }

  return (
    <HeaderMenuUser
      className={`${prefix}--bmrg-profile-settings-container`}
      src={src}
      userName={userName}
    >
      {({ closeModal }) => {
        return (
          <>
            <ModalHeader
              closeModal={() => {
                closeModal();
                setIsLoadingError(false);
                setIsSubmitError(false);
              }}
              title="User profile"
            />
            <ModalBody style={{ position: 'relative' }}>
              <p className={`${prefix}--bmrg-profile-settings__title`}>
                More user profile settings will be here eventually, but for now you can choose which
                Teams are shown in your sidebar in Launchpad.
              </p>
              <h2 className={`${prefix}--bmrg-profile-settings__subtitle`}>
                Teams visible in Launchpad sidebar
              </h2>
              <p className={`${prefix}--bmrg-profile-settings__description`}>
                Choose Teams to show or hide in your Launchpad sidebar and Catalog (useful for
                sensitive demos). You will not be able to access or view unchecked Teams from the
                sidebar, and cannot add items to them from Catalog.
              </p>
              {isLoading ? (
                <StructuredListSkeleton />
              ) : isLoadingError ? (
                <ErrorMessage style={{ color: '#F2F4F8' }} />
              ) : (
                <StructuredListWrapper className={`${prefix}--bmrg-profile-settings-list`}>
                  <StructuredListHead>
                    <StructuredListRow head>
                      <StructuredListCell head>
                        <Checkbox
                          checked={allTeamsAreChecked}
                          labelText={'Team Name'}
                          id={'team-name-batch-toggle'}
                          indeterminate={someTeamsAreChecked}
                          onChange={batchChangeTeamVisibility}
                        />
                      </StructuredListCell>
                    </StructuredListRow>
                  </StructuredListHead>
                  <StructuredListBody>
                    {sortBy(teams, 'name').map(({ name, id, visible }) => (
                      <StructuredListRow key={id}>
                        <StructuredListCell>
                          <Checkbox
                            checked={visible}
                            labelText={name}
                            id={id}
                            onChange={(isChecked, id) => {
                              return handleUpdateTeamVisibility(id, isChecked);
                            }}
                          />
                        </StructuredListCell>
                      </StructuredListRow>
                    ))}
                  </StructuredListBody>
                </StructuredListWrapper>
              )}
            </ModalBody>
            <ModalFooter>
              <Button kind="secondary" onClick={closeModal}>
                Cancel
              </Button>
              <Button
                kind="primary"
                type="submit"
                onClick={(e) => {
                  e.preventDefault();
                  handleSubmit({ closeModal });
                }}
              >
                {isSubmitError ? 'Try Again' : isSubmitting ? 'Saving...' : 'Save'}
              </Button>
            </ModalFooter>
          </>
        );
      }}
    </HeaderMenuUser>
  );
}

export default ProfileSettings;
