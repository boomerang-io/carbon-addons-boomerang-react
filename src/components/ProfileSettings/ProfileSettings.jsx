import React, { useCallback, useState, useEffect } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import {
  Button,
  Checkbox,
  InlineNotification,
  ModalHeader,
  ModalBody,
  ModalFooter,
  StructuredListWrapper,
  StructuredListHead,
  StructuredListBody,
  StructuredListRow,
  StructuredListCell,
  StructuredListSkeleton,
} from "@carbon/react";
import Avatar from "../Avatar";
import ErrorMessage from "../ErrorMessage";
import HeaderMenuUser from "../HeaderMenuUser";
import notify from "../Notifications/notify";
import ToastNotification from "../Notifications/ToastNotification";
import { prefix } from "../../internal/settings";
import sortBy from "lodash.sortby";

ProfileSettings.propTypes = {
  baseServiceUrl: PropTypes.string.isRequired,
  src: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
};

function ProfileSettings({ baseServiceUrl, src, userName }) {
  const [initialTeams, setInitialTeams] = useState([]);
  const [teams, setTeams] = useState([]);
  const [disableModal, setDisableModal] = useState([]);
  const [isLoading, setIsLoading] = useState();
  const [isSubmitting, setIsSubmitting] = useState();
  const [isLoadingError, setIsLoadingError] = useState();
  const [isSubmitError, setIsSubmitError] = useState();

  const fetchTeams = useCallback(
    ({ showLoading }) => {
      if (showLoading) {
        setIsLoading(true);
      }
      axios(`${baseServiceUrl}/launchpad/user`)
        .then((response) => {
          const teams = response.data.lowerLevelGroups;
          setTeams(teams);
          setDisableModal(response.data.lowerLevelGroups === undefined);
          setInitialTeams(teams);
        })
        .catch((err) => setIsLoadingError(err))
        .then(() => {
          setIsLoading(false);
        });
    },
    [baseServiceUrl]
  );

  useEffect(() => {
    fetchTeams({ showLoading: true });
  }, [fetchTeams]);

  function handleCloseModal({ closeModal }) {
    setIsLoadingError(false);
    setIsSubmitError(false);
    closeModal();
    fetchTeams({ showLoading: false });
  }

  function handleSubmit({ closeModal }) {
    setIsSubmitting(true);
    setIsSubmitError(false);
    axios
      .patch(`${baseServiceUrl}/users/profile`, {
        lowerLevelGroups: teams,
      })
      .then(() => {
        setIsSubmitting(false);
        notify(
          <ToastNotification subtitle="Successfully updated user settings" title="Update Settings" kind="success" />,
          { containerId: `${prefix}--bmrg-header-notifications` }
        );
        handleCloseModal({ closeModal });
      })
      .catch((err) => {
        setIsSubmitting(false);
        setIsSubmitError(true);
      });
  }

  const visibleTeamCount = teams?.filter((team) => team?.visible)?.length ?? 0;
  const allTeamsAreChecked = teams?.length === visibleTeamCount;
  const someTeamsAreChecked = visibleTeamCount > 0 && !allTeamsAreChecked;
  const isConfigDifferent = determineIfConfigIsDifferent(teams, initialTeams);

  function batchChangeTeamVisibility(action) {
    const updatedTeams = teams.map((team) => ({ ...team, visible: !allTeamsAreChecked }));
    setTeams(updatedTeams);
  }

  function determineIfConfigIsDifferent(teams, initialTeams) {
    let isConfigDifferent = false;
    for (let idx = 0; idx < teams?.length; idx++) {
      if (teams[idx]?.visible !== initialTeams[idx]?.visible) {
        isConfigDifferent = true;
        break;
      }
    }
    return isConfigDifferent;
  }

  function handleUpdateTeamVisibility(id, isChecked) {
    const updatedTeams = [];
    for (let team of teams) {
      const newTeam = { ...team };
      if (newTeam.id === id) {
        newTeam.visible = isChecked;
      }
      updatedTeams.push(newTeam);
    }
    setTeams(updatedTeams);
  }

  if (disableModal)
    return (
      <div className={`${prefix}--bmrg-profile-menu-user`}>
        <Avatar size="medium" src={src} userName={userName} />
        <p className={`${prefix}--bmrg-profile-menu-user__name`}> {userName ? userName : ""} </p>
      </div>
    );

  return (
    <HeaderMenuUser className={`${prefix}--bmrg-profile-settings-container`} src={src} userName={userName}>
      {({ closeModal }) => {
        return (
          <>
            <ModalHeader closeModal={() => handleCloseModal({ closeModal })} title={`User profile - ${userName}`} />
            <ModalBody style={{ maxHeight: "31.5rem" }}>
              <p className={`${prefix}--bmrg-profile-settings__title`}>
                More user profile settings will be here eventually, but for now you can choose which Teams are shown in
                your sidebar in Launchpad.
              </p>
              <h2 className={`${prefix}--bmrg-profile-settings__subtitle`}>Teams visible in Launchpad sidebar</h2>
              <p className={`${prefix}--bmrg-profile-settings__description`}>
                Choose Teams to show or hide in your Launchpad sidebar and Catalog (useful for sensitive demos). You
                will not be able to access or view unchecked Teams from the sidebar, and cannot add items to them from
                Catalog.
              </p>
              {isLoading ? (
                <StructuredListSkeleton />
              ) : isLoadingError ? (
                <ErrorMessage style={{ color: "#F2F4F8" }} />
              ) : teams?.length > 0 ? (
                <StructuredListWrapper className={`${prefix}--bmrg-profile-settings-list`}>
                  <StructuredListHead>
                    <StructuredListRow head>
                      <StructuredListCell head>
                        <Checkbox
                          checked={allTeamsAreChecked}
                          labelText={"Team Name"}
                          id={"team-name-batch-toggle"}
                          indeterminate={someTeamsAreChecked}
                          onChange={batchChangeTeamVisibility}
                        />
                      </StructuredListCell>
                    </StructuredListRow>
                  </StructuredListHead>
                  <StructuredListBody>
                    {sortBy(teams, "name").map(({ name, id, visible }) => (
                      <StructuredListRow
                        key={id}
                        className={!visible && `${prefix}--bmrg-profile-settings-list__row--disabled`}
                      >
                        <StructuredListCell>
                          <Checkbox
                            checked={visible}
                            labelText={name}
                            id={id}
                            onChange={(isChecked, id) => {
                              return handleUpdateTeamVisibility(id, isChecked);
                            }}
                          />
                        </StructuredListCell>
                      </StructuredListRow>
                    ))}
                  </StructuredListBody>
                </StructuredListWrapper>
              ) : (
                <p style={{ marginTop: "3rem", color: "#F2F4F8" }}>
                  No teams to configure. Join or create teams in Launchpad!
                </p>
              )}
            </ModalBody>
            <ModalFooter>
              {isSubmitError && (
                <div
                  style={{
                    position: "absolute",
                    top: "-5rem",
                    left: "2rem",
                    width: "90%",
                  }}
                >
                  <InlineNotification kind="error" title="Something's Wrong" subtitle="Failed to update user profile" />
                </div>
              )}
              <Button kind="secondary" onClick={() => handleCloseModal({ closeModal })}>
                Cancel
              </Button>
              <Button
                disabled={!isConfigDifferent || isSubmitting}
                kind="primary"
                type="submit"
                onClick={(e) => {
                  e.preventDefault();
                  handleSubmit({ closeModal });
                }}
              >
                {isSubmitError ? "Try Again" : isSubmitting ? "Saving..." : "Save changes"}
              </Button>
            </ModalFooter>
          </>
        );
      }}
    </HeaderMenuUser>
  );
}

export default ProfileSettings;
