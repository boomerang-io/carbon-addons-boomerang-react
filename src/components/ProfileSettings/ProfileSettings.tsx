import React, { useState, useEffect } from "react";
import { useQuery, useMutation, useQueryClient } from "react-query";
import {
  Button,
  Checkbox,
  ComposedModal,
  InlineNotification,
  ModalHeader,
  ModalBody,
  ModalFooter,
  StructuredListWrapper,
  StructuredListHead,
  StructuredListBody,
  StructuredListRow,
  StructuredListCell,
  StructuredListSkeleton,
} from "@carbon/react";
import Avatar from "../Avatar";
import ErrorMessage from "../ErrorMessage";
import HeaderMenuItem from "../Header/HeaderMenuItem";
import notify from "../Notifications/notify";
import ToastNotification from "../Notifications/ToastNotification";
import { serviceUrl, resolver } from "../../config/servicesConfig";
import sortBy from "lodash.sortby";
import { prefix } from "../../internal/settings";
import { LowerLevelGroup, User } from "../../types";

function determineIfConfigIsDifferent(teams: LowerLevelGroup[], initialTeams: LowerLevelGroup[]) {
  let isConfigDifferent = false;
  for (let idx = 0; idx < teams?.length; idx++) {
    if (teams[idx]?.visible !== initialTeams[idx]?.visible) {
      isConfigDifferent = true;
      break;
    }
  }
  return isConfigDifferent;
}

type Props = {
  closeModal: () => void;
  isOpen: boolean;
  baseServiceUrl: string;
  src: string;
  userName?: string;
};

function ProfileSettings({ baseServiceUrl, src, userName, isOpen, closeModal }: Props) {
  const queryClient = useQueryClient();

  const [initialTeams, setInitialTeams] = useState<LowerLevelGroup[]>([]);
  const [teams, setTeams] = useState<LowerLevelGroup[]>([]);

  const userUrl = serviceUrl.getLaunchpadUser({ baseServiceUrl });
  const profileUrl = serviceUrl.resourceUserProfile({ baseServiceUrl });

  const {
    data: user,
    isLoading: userIsLoading,
    error: userError,
  } = useQuery<User>({
    queryKey: userUrl,
    queryFn: resolver.query(userUrl),
  });

  const {
    mutateAsync: mutateUserProfile,
    isLoading: mutateUserProfileIsLoading,
    error: mutateUserProfileError,
  } = useMutation(resolver.patchUserProfile, {
    onSuccess: () => {
      queryClient.invalidateQueries(userUrl);
      queryClient.invalidateQueries(profileUrl);
    },
  });

  const disableModal = user?.lowerLevelGroups === undefined;

  useEffect(() => {
    const teams = user?.lowerLevelGroups ?? [];
    setInitialTeams(teams);
    setTeams(teams);
  }, [user]);

  async function handleSubmit({ closeModal }: any) {
    const body = {
      lowerLevelGroups: teams,
    };

    try {
      await mutateUserProfile({ baseServiceUrl, body });
      notify(
        <ToastNotification subtitle="Successfully updated user settings" title="Update Settings" kind="success" />,
        { containerId: `${prefix}--bmrg-header-notifications` }
      );
      closeModal();
    } catch (e) {
      // noop
    }
  }

  const visibleTeamCount = teams?.filter((team) => team?.visible)?.length ?? 0;
  const allTeamsAreChecked = teams?.length === visibleTeamCount;
  const someTeamsAreChecked = visibleTeamCount > 0 && !allTeamsAreChecked;
  const isConfigDifferent = determineIfConfigIsDifferent(teams, initialTeams);

  function batchChangeTeamVisibility() {
    const updatedTeams = teams.map((team) => ({ ...team, visible: !allTeamsAreChecked }));
    setTeams(updatedTeams);
  }

  function handleUpdateTeamVisibility(id: string, checked: boolean) {
    const updatedTeams = [];
    for (let team of teams) {
      const newTeam = { ...team };
      if (newTeam.id === id) {
        newTeam.visible = checked;
      }
      updatedTeams.push(newTeam);
    }
    setTeams(updatedTeams);
  }

  if (disableModal) {
    return (
      <div className={`${prefix}--bmrg-profile-menu-user`}>
        <Avatar size="medium" src={src} userName={userName} />
        <p className={`${prefix}--bmrg-profile-menu-user__name`}> {userName ? userName : ""} </p>
      </div>
    );
  }

  return (
    <ComposedModal
    aria-label="Profile Settings"
      open={isOpen}
      className={`${prefix}--bmrg-profile-settings-container ${prefix}--bmrg-header-modal`}
      onClose={closeModal}
    >
      <ModalHeader closeModal={closeModal} title={`User profile - ${userName}`} />
      <ModalBody style={{ maxHeight: "31.5rem" }}>
        <p className={`${prefix}--bmrg-profile-settings__title`}>
          More user profile settings will be here eventually, but for now you can choose which Teams are shown in your
          sidebar in Launchpad.
        </p>
        <h2 className={`${prefix}--bmrg-profile-settings__subtitle`}>Teams visible in Launchpad sidebar</h2>
        <p className={`${prefix}--bmrg-profile-settings__description`}>
          Choose Teams to show or hide in your Launchpad sidebar and Catalog (useful for sensitive demos). You will not
          be able to access or view unchecked Teams from the sidebar, and cannot add items to them from Catalog.
        </p>
        {userIsLoading ? (
          <StructuredListSkeleton />
        ) : userError ? (
          <ErrorMessage style={{ color: "#F2F4F8", padding: "1rem" }} />
        ) : teams?.length > 0 ? (
          <StructuredListWrapper className={`${prefix}--bmrg-profile-settings-list`}>
            <StructuredListHead>
              <StructuredListRow head>
                <StructuredListCell head>
                  <Checkbox
                    checked={allTeamsAreChecked}
                    id={"team-name-batch-toggle"}
                    indeterminate={someTeamsAreChecked}
                    labelText={"Team Name"}
                    onChange={batchChangeTeamVisibility}
                  />
                </StructuredListCell>
              </StructuredListRow>
            </StructuredListHead>
            <StructuredListBody>
              {(sortBy(teams, "name") as LowerLevelGroup[]).map(({ name, id, visible }) => (
                <StructuredListRow
                  key={id}
                  className={!visible ? `${prefix}--bmrg-profile-settings-list__row--disabled` : ""}
                >
                  <StructuredListCell>
                    <Checkbox
                      checked={visible}
                      id={id}
                      labelText={name}
                      onChange={(_: any, { checked, id }: { checked: boolean; id: string }) => {
                        return handleUpdateTeamVisibility(id, checked);
                      }}
                    />
                  </StructuredListCell>
                </StructuredListRow>
              ))}
            </StructuredListBody>
          </StructuredListWrapper>
        ) : (
          <p style={{ marginTop: "3rem", color: "#F2F4F8" }}>
            No teams to configure. Join or create teams in Launchpad!
          </p>
        )}
      </ModalBody>
      <ModalFooter>
        {mutateUserProfileError && (
          <div
            style={{
              position: "absolute",
              top: "-5rem",
              left: "2rem",
              width: "90%",
            }}
          >
            <InlineNotification kind="error" title="Something's Wrong" subtitle="Failed to update user profile" />
          </div>
        )}
        <Button data-modal-primary-focus kind="secondary" onClick={closeModal}>
          Cancel
        </Button>
        <Button
          disabled={!isConfigDifferent || mutateUserProfileIsLoading}
          kind="primary"
          type="submit"
          onClick={(e: any) => {
            e.preventDefault();
            handleSubmit({ closeModal });
          }}
        >
          {mutateUserProfileError ? "Try Again" : mutateUserProfileIsLoading ? "Saving..." : "Save changes"}
        </Button>
      </ModalFooter>
    </ComposedModal>
  );
}

export default ProfileSettings;

function ProfileSettingsMenuItem(props: Omit<Props, "isOpen" | "closeModal">) {
  const menuItemRef = React.useRef<HTMLButtonElement>(null);
  const [isOpen, setIsOpen] = React.useState(false);

  const handleClose = () => {
    setIsOpen(false);
    menuItemRef.current?.focus();
  };

  return (
    <>
      <HeaderMenuItem type="user" onClick={() => setIsOpen(!isOpen)} src={props.src} userName={props.userName} />
      <ProfileSettings isOpen={isOpen} closeModal={handleClose} {...props} />
    </>
  );
}

export { ProfileSettingsMenuItem };
