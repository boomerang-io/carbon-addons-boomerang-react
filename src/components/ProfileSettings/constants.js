export const PROFILE_SETTINGS_DATA = {
  id: '5e7377c6a97b78000125ed7c',
  email: 'trbula@us.ibm.com',
  name: 'Tim Bula',
  isFirstVisit: false,
  type: 'admin',
  isShowHelp: true,
  firstLoginDate: '2020-03-19T13:46:46.168+00:00',
  lastLoginDate: '2020-11-23T14:59:47.076+00:00',
  lowerLevelGroups: [
    { id: '5e73782ee04d700001a0024d', visible: true, name: 'Team 1' },
    { id: '5e739e7410d67a0001df192b', visible: false, name: 'Team 2' },
    { id: '5e7cfbe9be354b00015297b3', visible: true, name: 'Team 3' },
    { id: '5e7e4443b40af00001fc75c9', visible: false, name: 'Team 4' },
    { id: '5eab3e95333e56000137914b', visible: true, name: 'Team 5' },
    { id: '5e7d0297be354b00015298ba', visible: false, name: 'Team 6' },
    { id: '5f682bc0177fe668f9b3221f', visible: true, name: 'Team 7' },
    { id: '5f5a4a52f4f24e6354684b59', visible: false, name: 'Team 8' },
    { id: '5e9454692f5c4600015deda1', visible: true, name: 'Team 9' },
    { id: '5f6201fc5fc5376dc5487be0', visible: false, name: 'Team a' },
    { id: '5e8719e8b40af00001fc9acc', visible: true, name: 'Team b' },
    { id: '5f3ae8f4b7192225d8f4f0a5', visible: false, name: 'Team c' },
    { id: '5f3ae8f4b7192225d8f4f0a4', visible: false, name: 'Team d' },
    { id: '5f3ae8f4b7192225d8f4f0a3', visible: false, name: 'Team e' },
  ],
  pinnedToolIds: [],
  favoritePackages: [],
  personalizations: {},
  notificationSettings: {},
  status: 'active',
  teams: [],
};
