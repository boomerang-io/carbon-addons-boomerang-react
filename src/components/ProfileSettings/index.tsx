export { default, ProfileSettingsMenuItem } from "./ProfileSettings";
