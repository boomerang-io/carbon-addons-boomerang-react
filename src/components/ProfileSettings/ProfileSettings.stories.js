import React from 'react';
import { QueryClient, QueryClientProvider } from "react-query";
import axios from 'axios';
import ProfileSettings from './ProfileSettings';
import MockAdapter from 'axios-mock-adapter';
import { PROFILE_SETTINGS_DATA } from './constants';

const queryClient = new QueryClient({
  defaultOptions: { queries: { refetchOnWindowFocus: false }, mutations: { throwOnError: true } },
});

export default {
  title: 'ProfileSettings',
};

export const Default = () => {
  const mock = new MockAdapter(axios);
  mock.onGet('https://ibm.com/launchpad/user').reply(200, PROFILE_SETTINGS_DATA);
  mock.onPatch('https://ibm.com/users/profile').reply(200);
  return (
    <QueryClientProvider client={queryClient}>
      <div style={{ width: '16rem' }}>
        <ProfileSettings
          baseServiceUrl="https://ibm.com"
          src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"
          userName="Boomerang Ada"
        />
      </div>
    </QueryClientProvider>
  );
};

Default.story = {
  name: 'default',
};
