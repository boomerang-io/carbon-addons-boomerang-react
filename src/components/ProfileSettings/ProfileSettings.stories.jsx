import axios from "axios";
import ProfileSettings from "./ProfileSettings";
import MockAdapter from "axios-mock-adapter";
import { PROFILE_SETTINGS_DATA } from "./constants";

const mock = new MockAdapter(axios);
mock.onGet("https://ibm.com/launchpad/user").reply(200, PROFILE_SETTINGS_DATA);
mock.onPatch("https://ibm.com/users/profile").reply(200);

export default {
  title: "Platform/ProfileSettings",
  component: ProfileSettings,
};

export const Default = () => {
  return (
    <div style={{ width: "16rem", background: "#061727" }}>
      <ProfileSettings
        baseServiceUrl="https://ibm.com"
        src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"
        userName="Boomerang Ada"
      />
    </div>
  );
};
