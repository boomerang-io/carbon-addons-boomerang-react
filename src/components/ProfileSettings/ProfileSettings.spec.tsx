import { expect, test } from "vitest";
import { QueryClient, QueryClientProvider } from "react-query";
import { render, fireEvent, waitFor } from "@testing-library/react";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
// @ts-expect-error TS(7016): Could not find a declaration file for module 'jest... Remove this comment to see the full error message
import { axe } from "jest-axe";
import ProfileSettings from "./ProfileSettings";
import { PROFILE_SETTINGS_DATA } from "./constants";

const baseServiceUrl = "http://boomerang.com";

const queryClient = new QueryClient({
  // @ts-expect-error TS(2322): Type '{ refetchOnWindowFocus: false; throwOnError:... Remove this comment to see the full error message
  defaultOptions: { queries: { refetchOnWindowFocus: false, throwOnError: false }, mutations: { throwOnError: true } },
});

test("Profile Settings success", async () => {
  const mock = new MockAdapter(axios);
  mock.onGet(`${baseServiceUrl}/launchpad/user`).reply(200, PROFILE_SETTINGS_DATA);
  mock.onPatch(`${baseServiceUrl}/users/profile`).reply(200);
  const { findByText, getByLabelText, getByText, queryByText } = render(
    <QueryClientProvider client={queryClient}>
      <ProfileSettings baseServiceUrl={baseServiceUrl} userName="Boomerang Joe" src="joe@ibm.com" />
    </QueryClientProvider>
  );
  const userBtn = getByText("Boomerang Joe");
  fireEvent.click(userBtn);
  const btn = await findByText(/Save changes/i);
  (expect(btn) as any).toBeDisabled();
  const allToggle = getByLabelText("Team Name");
  const team1Toggle = getByLabelText("Team 1");
  const team2Toggle = getByLabelText("Team 2");
  fireEvent.click(allToggle);
  (expect(team1Toggle) as any).toBeEnabled();
  (expect(team2Toggle) as any).toBeEnabled();
  (expect(btn) as any).toBeEnabled();
  fireEvent.click(btn);
  await waitFor(() =>
    expect(queryByText(/For any questions or concerns about business personal information/i)).toBeNull()
  );
});

test("Profile Settings error", async () => {
  const mock = new MockAdapter(axios);
  mock.onGet(`${baseServiceUrl}/launchpad/user`).reply(200, PROFILE_SETTINGS_DATA);
  mock.onPatch(`${baseServiceUrl}/users/profile`).networkError();
  const { findByText, findByLabelText, getByText } = render(
    <QueryClientProvider client={queryClient}>
      <ProfileSettings baseServiceUrl={baseServiceUrl} userName="Boomerang Joe" src="joe@ibm.com" />
    </QueryClientProvider>
  );
  const userBtn = getByText("Boomerang Joe");
  fireEvent.click(userBtn);
  const allToggle = await findByLabelText(/Team Name/i);
  fireEvent.click(allToggle);
  const btn = await findByText(/Save changes/i);
  fireEvent.click(btn);
  await waitFor(() => (expect(getByText(/try again/i)) as any).toBeInTheDocument());
});

test("Profile Settings accessibility", async () => {
  const { container } = render(
    <QueryClientProvider client={queryClient}>
      <ProfileSettings baseServiceUrl={baseServiceUrl} userName="Boomerang Joe" src="joe@ibm.com" />
    </QueryClientProvider>
  );
  const results = await axe(container);
  (expect(results) as any).toHaveNoViolations();
});
