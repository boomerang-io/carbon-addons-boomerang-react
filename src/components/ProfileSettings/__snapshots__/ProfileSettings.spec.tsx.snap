// Vitest Snapshot v1

exports[`Profile Settings > snapshot 1`] = `
<body
  class="cds--body--with-modal-open"
>
  <div>
    <div
      aria-hidden="false"
      class="cds--modal is-visible cds--bmrg-profile-settings-container cds--bmrg-header-modal"
      role="presentation"
      tabindex="-1"
    >
      <div
        aria-label="Profile Settings"
        aria-modal="true"
        class="cds--modal-container"
        role="dialog"
      >
        <button
          class="cds--visually-hidden"
          type="button"
        >
          Focus sentinel
        </button>
        <div
          class="cds--modal-container-body"
        >
          <div
            class="cds--modal-header"
          >
            <h3
              class="cds--modal-header__heading cds--type-beta"
            >
              User profile - Boomerang Joe
            </h3>
            <button
              aria-label="Close"
              class="cds--modal-close"
              title="Close"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="cds--modal-close__icon"
                fill="currentColor"
                focusable="false"
                height="20"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                />
              </svg>
            </button>
          </div>
          <div
            class="cds--modal-content"
            style="max-height: 31.5rem;"
          >
            <p
              class="cds--bmrg-profile-settings__title"
            >
              More user profile settings will be here eventually, but for now you can choose which Teams are shown in your sidebar in Launchpad.
            </p>
            <h2
              class="cds--bmrg-profile-settings__subtitle"
            >
              Teams visible in Launchpad sidebar
            </h2>
            <p
              class="cds--bmrg-profile-settings__description"
            >
              Choose Teams to show or hide in your Launchpad sidebar and Catalog (useful for sensitive demos). You will not be able to access or view unchecked Teams from the sidebar, and cannot add items to them from Catalog.
            </p>
            <div
              class="cds--skeleton cds--structured-list"
            >
              <div
                class="cds--structured-list-thead"
              >
                <div
                  class="cds--structured-list-row cds--structured-list-row--header-row"
                >
                  <div
                    class="cds--structured-list-th"
                  >
                    <span />
                  </div>
                  <div
                    class="cds--structured-list-th"
                  >
                    <span />
                  </div>
                  <div
                    class="cds--structured-list-th"
                  >
                    <span />
                  </div>
                </div>
              </div>
              <div
                class="cds--structured-list-tbody"
              >
                <div
                  class="cds--structured-list-row"
                >
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                </div>
                <div
                  class="cds--structured-list-row"
                >
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                </div>
                <div
                  class="cds--structured-list-row"
                >
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                </div>
                <div
                  class="cds--structured-list-row"
                >
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                </div>
                <div
                  class="cds--structured-list-row"
                >
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                  <div
                    class="cds--structured-list-td"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="cds--modal-footer cds--btn-set"
          >
            <button
              class="cds--btn cds--btn--secondary"
              data-modal-primary-focus="true"
              tabindex="0"
              type="button"
            >
              Cancel
            </button>
            <button
              class="cds--btn cds--btn--primary cds--btn--disabled"
              disabled=""
              tabindex="0"
              type="submit"
            >
              Save changes
            </button>
          </div>
        </div>
        <button
          class="cds--visually-hidden"
          type="button"
        >
          Focus sentinel
        </button>
      </div>
    </div>
  </div>
</body>
`;
